lxc/completion.go

```diff
@@ -10,81 +11,204 @@ import (
+		results = append(results, remotes...)
+		directive |= directives
 	}
 
-	results := make([]string, 0, len(groups))
+	return results, directive
+}
 
-	for _, group := range groups {
-		var name string
+// cmpTopLevelResourceInRemote returns completions for a given entity type in a given remote, based on the partial `toComplete` argument.
+func (g *cmdGlobal) cmpTopLevelResourceInRemote(remote string, entityType string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	instanceServerResourceNameFunc, ok := topLevelInstanceServerResourceNameFuncs[entityType]
+	if !ok {
+		imageServerResourceNameFunc, ok := topLevelImageServerResourceNameFuncs[entityType]
+		if !ok {
+			return handleCompletionError(fmt.Errorf("Cannot compare names: Entity type %q is not a top level resource", entityType))
+		}
 
-		if resource.remote == g.conf.DefaultRemote && !strings.Contains(toComplete, g.conf.DefaultRemote) {
-			name = group
-		} else {
-			name = resource.remote + ":" + group
+		server, err := g.conf.GetImageServer(remote)
+		if err != nil {
+			return handleCompletionError(err)
 		}
 
-		results = append(results, name)
+		names, err := imageServerResourceNameFunc(server)
+		if err != nil {
+			handleCompletionError(err)
+		}
+
+		return completionsFor(names, "", toComplete), cobra.ShellCompDirectiveNoFileComp
 	}
 
-	if !strings.Contains(toComplete, ":") {
-		remotes, directives := g.cmpRemotes(toComplete, false)
-		results = append(results, remotes...)
-		cmpDirectives |= directives
+	server, err := g.conf.GetInstanceServerWithAdditionalConnectionArgs(remote, &lxd.ConnectionArgs{SkipGetServer: true})
+	if err != nil {
+		return handleCompletionError(err)
 	}
 
-	return results, cmpDirectives
+	names, err := instanceServerResourceNameFunc(server)
+	if err != nil {
+		return handleCompletionError(err)
+	}
+
+	return completionsFor(names, "", toComplete), cobra.ShellCompDirectiveNoFileComp
+}
+
+// configOptionAppender returns two functions, the first appends config options to a slice, the second returns the current
+// slice value. Config options are trimmed to the last '.' to improve discoverability. If there is no trailing '.', the
+// given suffix will be appended. This allows lists like: ["gid=", "id=", "mdev=", "mig."] when setting configuration.
+func configOptionAppender(toComplete string, suffix string, size int) (func(option string), func() []string) {
+	var out []string
+	if size > 0 {
+		out = make([]string, 0, size)
+	}
+
+	return func(option string) {
+			trimmed, found := strings.CutPrefix(option, toComplete)
+			if !found {
+				return // Does not have prefix
+			}
+
+			var key string
+			remaining, _, ok := strings.Cut(trimmed, ".")
+			if ok {
+				// The option contains a `.<something>` after the completion, append only up to the '.'
+				key = toComplete + remaining + "."
+			} else {
+				// The option does not contain any more `.<something>'s after the completion, append the suffix.
+				key = option + suffix
+			}
```

Comment:
```suggestion
			var key string
			before, after, ok := strings.Cut(trimmed, ".")
			if ok {
				// The option contains a `.<something>` after the completion, append only up to the '.'
				key = toComplete + remaining + "."
			} else if after == "*" {
			        // The option ends with "*", don't append suffix.
			        key = option
			} else {
				// The option does not contain any more `.<something>'s after the completion, append the suffix.
				key = option + suffix
			}
```

We shouldn't append the suffix for keys such as `linux.sysctl.`, or else we end up completing `linux.sysctl.=`.

Reply:
Fixed.